Modify code to follow hexagonal architecture

Dysproz opened the pull request:
This PR refers to issue #2.

It changes code architecture to use hexagonal architecture following patterns described in [this](https://medium.com/@matiasvarela/hexagonal-architecture-in-go-cfd4e436faa3) article.

Architecture describes domain types in `internal/core/domain` package and applies them to services defined in `internal/core/services`. However, dependencies between services are not defined directly, but using interfaces defined in `internal/core/ports` - that way it's possible to easily substitute service with a different one without affecting other services.

`internal/repositories` contain code with drivers for 3rd party services - in this case it's a mongoDB database.
`internal/handlers` contain code for handling HTTP requests used here with clientAPI.

To fit into this code structure I had to make a bit tricky split of .proto file into **domain.proto** and **portsprotcol.proto** and generate then separately. The reason is that definitions of `messages` are in **domain.proto** and generate Go code in *domain* package under `internal/core/domain` while `services` are in **portsprotocol.proto** and Generate Go code under `internal/core/services/portprotocol` which refers to types defined in domain part. 

This PR changes only code structure and does not change any other logic in code. Further issues will be addressed in further PRs.

fabriziomoscon commented on internal/core/ports/services.go:
```diff
@@ -0,0 +1,18 @@
+	Watch() <-chan domain.Entry
+}
+
+type GRPCClientService interface {
```
GRPC is a technology specific, so it cannot stay in core. Image that the core contains business logic used by both a server and a command line application, the cli doesn't use GRPC.
Another way to think about this is a service that implements both grpc and http transport.

fabriziomoscon commented:
I think you had a good initial take to this, but it is not fully working as hexagonal because the core imports not core packages.
The crucial point is that the core package should only contain business logic without any vendor package and without importing any of your packages.
Your implementation would benefit by removing the excessive packaging.
If you think about the problem we have only a few concerns:
- business logic, which should be in charge of modelling the data in memory and validating for this application
- transport, in your case protobuf over GPRC
- repository, data access, can be anything, file, map, postgres, mongo, another grpc service

if you flatten the packages you can use file names to document what the file contains and what should do. No reason to further separate inside core package in micro packages.

the ports/adapter are your interface and you should interface the functions of the repository and the function of the business logic.

In spirit of sharing ideas you might find this video interesting: https://www.youtube.com/watch?v=MzTcsI6tn-0
when I was learning about this concepts, I found this repo very useful: https://github.com/tensor-programming/hex-microservice/tree/part-3

We do appreciate your efforts to improve your test.

Dysproz commented:
@fabriziomoscon 
I've moved grpc client and server logic to handlers as you are right that it's a technology specific part and also handles connections similarly as REST API handles HTTP connection. I don't know why I didn't see that before.
I think that it rectified a lot of in the core code. Now main functions import only from the project or some basic libraries and not from '3rd party' grpc library.
Also I've moved all the steps before creating actual client or server that were in main to NewPortsProtocolServer and NewGrpcClient functions which made a main functions cleaner and easier to read.

I hope that now it's a truly hexagonal architecture. :) 
 

P.S. 
I've merged here all the other PRs to make it easier to work on the code and not worry about rebasing and keeping all the PRs consistent with each other.

fabriziomoscon commented on internal/core/services/jsonparse/service.go:
```diff
@@ -0,0 +1,65 @@
+package jsonparser
```
if you take this package outside of `core`, for example in `internal/datainput` file name `jsonparser.go`, your design will allow to extend with other kind of data input, since the data input is not part of the core of the hexagon.

fabriziomoscon commented on internal/core/services/portsprotocol/portsprotocol_fake.go:
```diff
@@ -0,0 +1,90 @@
+package portsprotocol
```
this is not core. Is simply a transport implementation, what if your app in the future will accept both gRPC and HTTP transport? This should be into package `internal/transport`

fabriziomoscon commented on internal/handlers/grpcsrv.go:
```diff
@@ -0,0 +1,71 @@
+package handlers
```
why is this package named handlers? This should be package `internal/grpc` to allow a generic design. Note that gRPC handlers are typically referred to as resolvers not handlers.

fabriziomoscon commented:
@Dysproz this is much better than before. Thank you for all the time you spent with this + the learning.
From my side it is good enough to progress to the interview.

Further improvements:
- your project will be further simplified and easier to maintain if you remove even more folders/packages. You should really only have 3 packages: core, repository and grpc or transport. The extra folders inside are actually adding bad complexity as open opportunity for circular reference and other issues.
- note that at 90poe we do not allow git merge, only rebase, so I suggest for the interview you familiarise with it.

fabriziomoscon commented on internal/handlers/restapi_test.go:
```diff
@@ -0,0 +1,103 @@
+		jsonparser.NewStream(),
+	)
+	for _, tc := range tests {
+		var jsonPayload = []byte(fmt.Sprintf(`{"key":"%v"}`, tc.key))
```
note that this can be improved for parallelism with sub tests https://dave.cheney.net/2019/05/07/prefer-table-driven-tests

Dysproz committed:
feat: heaxagonal architecture, context use, code readability, makefile, table tests, errors and exit handling

Signed-off-by: Dysproz <krasuski.szymon.piotr@gmail.com>
